### PR TITLE
Refactored PGoApi to make it immutable, PGoApiRequest introduced (solve missing pokemon)

### PR DESCRIPTION
**Pull request by DeirhX**
This is a solution to concurrency problems with PGoApi as mentioned in https://github.com/AHAAAAAAA/PokemonGo-Map/issues/1897. PGoApi was split into immutable (login) and mutable parts (position, request collection) so it can be used safely between threads.
Mutable parts were transferred into PGoApiRequest class which now represent the state of an RPC request and each thread now gets a new instance of it to prevent data races as are happening now.

## Description
Spinned of PGoApiRequest from PGoApi. PGoApi is now immutable and thread-safe, representing  established login to server. PGoApiRequest is now instanced from PGoApi for each new RPC request you want to make.

Feel free to refactor further and adjust coding style if needed ;) 

## Motivation and Context
Described in depth in https://github.com/AHAAAAAAA/PokemonGo-Map/issues/1897. Data-races are now causing scan results to be incomplete and pokemon missing from scan results.

## How Has This Been Tested?
Ran a stress test and scans are now complete and faster than before.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.   // I hope ;)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by AHAAAAAAA** [comment]
To replace #1913 hotfix once #671 is merged in following API update.

...
Say that 10 times in a row.

**Comment by Chlodochar** [comment]
Best would be to do the PR on pgoapi. After #671 is merged, it's trivial to use the new pgoapi.

File: pogom/pgoapi/pgoapi.py [hunk 1]
```diff
@@ -38,142 +38,175 @@
...
         self._position_alt = 0
 
+    def get_auth_provider(self):
+        return self._auth_provider
+
+    def get_endpoint(self):
+        return self._api_endpoint
+
+    def get_position(self):
+        return (self._position_lat, self._position_lng, self._position_alt)
```
Comment (Chlodochar) [inline pogom/pgoapi/pgoapi.py]: Is `get_position` and position in general still needed in PGoApi? `set_position` is being called on the request before every call. That seems good. It seems passing the position from PGoApi to request isn't needed.

**Comment by CT9AJ** [comment]
Is there anyway I could apply this hotfix?

**Comment by Chlodochar** [comment]
From what I remember:
```
git fetch https://github.com/DeirhX/PokemonGo-Map dev-feature-lockfree && git merge FETCH_HEAD
```

**Comment by CT9AJ** [comment]
I am running the Master version, I assume that would not apply for me right?

**Comment by Chlodochar** [comment]
@CT9AJ You'll probably be better off waiting for this all to drop into master.

**Comment by CT9AJ** [comment]
I actually already just switched to the Dev build lol.. Where do I put that line you just gave me? Do I do a Git Bash Here from the Dev folder?

**Comment by Chlodochar** [comment]
Yes, but please leave these questions somewhere else. See discord. The pull requests are specific to feature additions that are usually (initially) unstable.

**Comment by CT9AJ** [comment]
Ok thanks.

**Comment by Josephvb10** [comment]
Hmm, seems like this could be introducing a problem. 

Tried to use this fix, but after a few hours, the infamous bug that causes Pokémon to stop appearing return. The solution is to restart the script manually.  @DeirhX

File: pogom/pgoapi/pgoapi.py (hunk 1 above)
Comment (DeirhX) [inline pogom/pgoapi/pgoapi.py]: Yeah, I was considering removing at least the getter. As for the position value itself I was unsure if the original login position wouldn't be needed for future requests but it isn't right now.

**Comment by DeirhX** [comment]
@Josephvb10 Interesting, I wouldn't expect it to be introducing new issues. If Pokemon stop appearing altogether that would be probably due to server issues as currently the PTC login is again unstable. Or do you mean you are getting incomplete results and Pokemon missing from some of the scanned cells?

**Comment by Josephvb10** [comment]
@DeirhX Sorry, I edited the post. Seems like the TPC login server has been down a lot for the last hour, so that could be the problem. I will keep this version running and report back if I found another issue.

**Comment by DeirhX** [comment]
Myself I have kept running it overnight and apart from scanners having problems with PTC servers, results seem stable for me still.

File: pogom/pgoapi/pgoapi.py (hunk 1 above)
Comment (DeirhX) [inline pogom/pgoapi/pgoapi.py]: Anyway I agree that the position can go to keep it simple. We can reintroduce it should future needs call for it.

**Comment by everix1992** [comment]
I pulled this down last night so I could test it as well. It worked flawlessly for an hour or two and then I ran into the same errors that @Josephvb10 did. Restarting the script seemed to fix the problem. I assumed it had something to do with the number of API requests that were being sent from a properly multithreaded server, but this is just wild speculation.

**Comment by Josephvb10** [comment]
Yes. Happened to me again. Looks like it wasn't related to the servers being down after all. @DeirhX

**Comment by Chlodochar** [comment]
It's quite hard to wrap my head around the code, but having `api` be global in `search.py` seems dangerous. https://github.com/DeirhX/PokemonGo-Map/blob/cf64dd3d1f1007926b09681cc4a8f2447a7de9bb/pogom/search.py#L153
Is this actually needed? I'm interested whether a login per thread (and completely separating state) will solve the issue. Or are multiple login sessions for the same account not allowed?

**Comment by DeirhX** [comment]
Since PGoApi is now immutable, sharing it as global is harmless as there is no way you can change its state (as long you synchronize all assignments to it). I don't think we need a separate login per thread so that's why it is shared - but feel free to suggest a better way to share it. What we do need is are separate PGoApiRequests per thread so constructing the RPC call in one thread doesn't overwrite values set from another thread which is what was happening now (since it is being shared and has internal state).

@everix1992 , @Josephvb10 this refactor doesn't attempt to address issues with failing the requests and subsequent relogin to PTC server. That's a different issue (however I have a potential solution ready that depends on this PR being merged in - but it's hard to say if it's 100% yet). This PR addresses the issue of pokemon being missing from some cells you successfully scanned because multiple threads overwrote the position data of each other's request and thus multiple requests used the same position (yet still landed in different cells since that's a different variable). It also addresses performance in cases where multiple threads concurrently wrote into PGoApi request and duplicated many of their requests.

**Comment by DeirhX** [comment]
To explain the reasoning behind founding PGoApiRequest class better:
Currently, all threads share the same PGoApi.position and PGoApi.req_method_list. This is inherently bad because requests are made by first updating the position variable then populating req_method_list and then calling call() to actually start dispatching the request. If another thread happens to also want to call at the same moment and starts preparing the request, it overwrites the values of the first thread because **there is only one global PGoApi object**.  Position and req_method_list must not be shared between threads! They represent unique state of each thread's requests and each thread wants to make a different request!

**Comment by DeirhX** [comment]
On the other hand login information is never modified after login is successfully established. Therefore this can be safely shared between threads without it requiring to be copied around. And since Niantic uses HTTPS as the transport layer, there is no connection, no open socket that would make a problem when accessed from multiple threads. So I think we can safely retain one login state per process - which is what we already do.

**Comment by DeirhX** [comment]
On the other hand login information is never modified after login is successfully established. Therefore this can be safely shared between threads without it requiring to be copied around. And since Niantic uses HTTPS as the transport layer, there is no connection, no open socket that would make a problem when accessed from multiple threads. So I think we can safely retain one login state per process - which is what we already do.  It's the mutable internal state of constructing additional requests that is a problem.

File: pogom/search.py [hunk 2]
```diff
@@ -149,21 +150,20 @@ def search(args, i):
...
         if remaining_time > 60:
             log.info("Skipping Pokemon Go login process since already logged in for another {:.2f} seconds".format(remaining_time))
         else:
-            login(args, position)
-    else:
-        login(args, position)
+            active_api = None
 
-    lock = Lock()
+    if not active_api:
```
Comment (Chlodochar) [inline pogom/search.py]: It seems safer to have one thread at once do `api = login(args, position)` and have others wait for that operation. At the moment multiple threads can login and set `api` at the same time. I'm not sure that always results in the right token. Why not add a global lock `apiLock` that is used like:

```
apiLock = Lock()
...
global apiLock
...
if not active_api:
    apiLock.acquire()
    try:
        # Check whether another thread already refreshed api while we were waiting for the lock.
        if api._auth_provider._ticket_expire / 1000 - time.time() < 0:
            api = login(args, position)
        active_api = api
    finally:
        apiLock.release()
```

Let me know if it's unnecessary.

Comment (DeirhX) [inline pogom/search.py]: You are completely right, I forgot the lock in this revision. All other threads should indeed wait whenever a login operation is already in progress, at least that was my intention as well (and current unpushed implementation). 

Thanks for the heads up. I'll commit the fix once I get home.

**Comment by DeirhX** [comment]
@Chlodochar Pushed new commit that resolves the missing lock and also removes redundant lock on search pass since it doesn't use any shared data anymore (except database commit which should take care of its own locking dependent on engine). Please check whether you agree with it being redundant. It should speed up the scan a little bit more.

Also moved the login code into the search worker, closer to actual request happening. This will give the search worker opportunity to request a relogin if it feels the connection has gone stale (wrong responses or no responses at all). Previously there was not any possibility to wait for the relogin from within inside the search. This should help with issues @everix1992 and @Josephvb10 were experiencing since relogin (done by restarting the process) seems to help in these cases. It is, however, a larger change so feel free to cherry pick only the solution to the previous problem.

**Comment by everix1992** [comment]
Running for an hour now and everything seems fine. I'll keep you updated on whether I have any issues.

**Comment by everix1992** [comment]
I'm not sure if I just got used to using your changes, but the search seems to run much slower on the main develop branch as well, so the sooner we can get this pulled in the better. Let me know if there is anything I can do to help with that! I may have a harder time testing it now though as it's become a pain switching between this and develop now (due to config and webhook changes).

**Comment by Chlodochar** [comment]
This looks good. We need to get this into pgoapi as well. (@tejado)

**Comment by tejado** [comment]
I have it already in my local codebase. Will push it today.

**Comment by tejado** [comment]
FYI: https://github.com/tejado/pgoapi/commit/1b355130ad6d0a64a3267eb9dc3186cbbcd6aa68

**Comment by AHAAAAAAA** [comment]
Work on this is moved to #3106

**Pull request closed without merging**